store/store.go

```diff
@@ -499,6 +504,60 @@ func (s *Store[H]) get(ctx context.Context, hash header.Hash) ([]byte, error) {
 	return data, nil
 }
 
+// try advance contiguous head based on already written headers.
+func (s *Store[H]) advanceContiguousHead(ctx context.Context, headers ...H) {
+	// always inform heightSub about new headers seen
+	for _, h := range headers {
+		s.heightSub.UnblockHeight(h.Height())
+	}
+
+	currHead := s.contiguousHead.Load()
+	if currHead == nil {
+		return
+	}
+	currHeight := (*currHead).Height()
+	prevHeight := currHeight
+
+	// TODO(cristaloleg): benchmark this timeout or make it dynamic.
+	advCtx, advCancel := context.WithTimeout(ctx, 10*time.Second)
+	defer advCancel()
+
+	var newHead H
+	for {
+		h, err := s.getByHeight(advCtx, currHeight+1)
+		if err != nil {
+			break
+		}
+		newHead = h
+		currHeight++
+	}
+
+	if currHeight > prevHeight {
+		s.updateContiguousHead(newHead, currHeight)
+	}
+}
+
+func (s *Store[H]) updateContiguousHead(newHead H, newHeight uint64) {
+	s.contiguousHead.Store(&newHead)
+	s.heightSub.UnblockHeight(newHeight)
+	log.Infow("new head", "height", newHead.Height(), "hash", newHead.Hash())
+	s.metrics.newHead(newHead.Height())
+
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancel()
+
+	b, err := newHead.Hash().MarshalJSON()
+	if err != nil {
+		log.Errorw("cannot marshal new head",
+			"height", newHead.Height(), "hash", newHead.Hash(), "err", err)
+	}
+
+	if err := s.ds.Put(ctx, headKey, b); err != nil {
```

Comment:
Why we decouple this from the parent context coming from `advanceContiguousHead`?

Reply:
From the previous thread I wanted to show why we need to do that. In short: if we entered `if currHeight > prevHeight {` we definitely want to update the `headKey` (even if parent ctx is done) to prevent desync.

Reply:
Ok, so you mean atomicity.

However, I don't get this entirely. We still have context and timeout that is gonna be canceled after 5 secs. If you remove it here completely and simply pass `context.Backgrount()`, then yeah, its gonna wait for as long as needed until the key is updated and achieve what you describe. However, if we set the independent context for 5 secs, it is effectively the same as having it with the parent one with additional 5 secs on it.

Reply:
I see your point now. I misread the code with break, thinking it does return there

Reply:
> BECAUSE parent context can be cancelled BUT we have something to update.

But if child context cancels on its own after 5 secs, what's the point? The following is still true 

> If you remove it here completely and simply pass context.Backgrount(), then yeah, its gonna wait for as long as needed until the key is updated and achieve what you describe.

Reply:
I guess to avoid complete deadlocking... Yeah

Reply:
Blocking forever is also not the best option 'cause it can take...eternity. So, 2 edge cases

Reply:
But if that happens, it means we are in the desynced state, which we will need to recover

Reply:
Actually, I think we have to try advancing on node start. Imagine that store terminated without completing advancing, like in the case above, which means on restart it's gonna report wrong `Head` until its Appended again, which is a bug imo

Reply:
I agree to desynced state 'cause `headKey` and headers flush are separate things. So, let's propagate context then and leave it as it's. 

Idea regarding Start is definitely for the next PR (if any).

Done.